src/Microsoft.AspNet.Mvc.DataAnnotations/DataAnnotationsModelValidator.cs

```diff
@@ -26,6 +27,115 @@ public DataAnnotationsModelValidator(ValidationAttribute attribute, IStringLocal
+                    {
+                        return new MinLengthAttributeAdapter((MinLengthAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(RegularExpressionAttribute), (modelMetadata) =>
+                    {
+                        return new RegularExpressionAttributeAdapter((RegularExpressionAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CompareAttribute), (modelMetadata) =>
+                    {
+                        return new CompareAttributeAdapter((CompareAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata, metadataProvider);
+
+                    }
+                },
+                {
+                    typeof(StringLengthAttribute), (modelMetadata) =>
+                    {
+                        return new StringLengthAttributeAdapter((StringLengthAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CreditCardAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "creditcard", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(EmailAddressAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "email", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(PhoneAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "phone", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(UrlAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "url", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                }
+            };
+        }
+
+        private string GetErrorMessage(ModelMetadata metadata, IModelMetadataProvider metadataProvider)
+        {
+            if (Attribute == null)
+            {
+                throw new ArgumentNullException(nameof(Attribute));
+            }
+
+
+            var attrType = Attribute.GetType();
+
+            var dict = BuildErrorMessageDict(metadataProvider);
```

Comment:
Dictionary contains no dynamic information at all. Might as well be stored in a `static` field.

On the extensibility I mentioned below, we're missing support for custom `ValidationAttribute` subclasses all up. That is `DataAnnotationsClientModelValidatorProvider` lacks anything similar to `RegisterAdapterFactory()` from MVC 5. This is a separate bug, important but not something to fix here. For this PR, `DataAnnotationsClientModelValidatorProvider` and `DataAnnotationsModelValidator` should share an adapter lookup table. Refactor existing code so it's built once and available here. We can make that shared table extensible later.